From 1e39c97eaff6341b309666fd155ae7ddb6709138 Mon Sep 17 00:00:00 2001
From: Shilin Victor <chrono.monochrome@gmail.com>
Date: Mon, 2 Jul 2018 16:21:58 +0300
Subject: [PATCH 38/40] healthd: Use N-style percent display for off-mode
 charger

The charger animation field handling has changed with O.
If there is no custom animation defined with a field font,
fall back to default system font for batt percent display.

Change-Id: I16b94dc8ac6c488322cbbb48054059bf91905bb8
---
 healthd/healthd_draw.cpp | 45 +++++++++++++++++++++++++++++++++++++--------
 1 file changed, 37 insertions(+), 8 deletions(-)

diff --git a/healthd/healthd_draw.cpp b/healthd/healthd_draw.cpp
index ea3d991c0..ff93f4bd8 100644
--- a/healthd/healthd_draw.cpp
+++ b/healthd/healthd_draw.cpp
@@ -158,17 +158,46 @@ void HealthdDraw::draw_percent(const animation* anim) {
   const animation::text_field& field = anim->text_percent;
   if (field.font == nullptr || field.font->char_width == 0 ||
       field.font->char_height == 0) {
-    return;
-  }
 
-  std::string str = base::StringPrintf("%d%%", cur_level);
+    char cap_str[5];
+    int x, y;
+    int str_len_px;
+    int batt_height = 0;
+    // get height of battery image to draw text below
+    const animation::frame& frame = anim->frames[anim->cur_frame];
+    if (anim->num_frames != 0) {
+        // nothing else should happen actually
+        batt_height = gr_get_height(frame.surface);
+    }
+
+    snprintf(cap_str, (5), "%d%%", cur_level);
+    str_len_px = gr_measure(gr_sys_font(), cap_str);
+    x = (gr_fb_width() - str_len_px) / 2;
+    // draw it below the battery image
+    y = (gr_fb_height() + batt_height) / 2 + char_height_ * 2;
+
+    if (cur_level < 15) {
+      gr_color(255, 0, 0, 255); // red
+    } else if (cur_level < 50) {
+      gr_color(255, 128, 0, 255); // orange
+    } else if (cur_level < 90) {
+      gr_color(255, 255, 0, 255); // yellow
+    } else { // cur_level >= 90
+      gr_color(0, 255, 0, 255); // green
+    }
+
+    draw_text(gr_sys_font(), x, y, cap_str);
 
-  int x, y;
-  determine_xy(field, str.size(), &x, &y);
+  } else {
+    std::string str = base::StringPrintf("%d%%", cur_level);
 
-  LOGV("drawing percent %s %d %d\n", str.c_str(), x, y);
-  gr_color(field.color_r, field.color_g, field.color_b, field.color_a);
-  draw_text(field.font, x, y, str.c_str());
+    int x, y;
+    determine_xy(field, str.size(), &x, &y);
+
+    LOGV("drawing percent %s %d %d\n", str.c_str(), x, y);
+    gr_color(field.color_r, field.color_g, field.color_b, field.color_a);
+    draw_text(field.font, x, y, str.c_str());
+  }
 }
 
 void HealthdDraw::draw_battery(const animation* anim) {
-- 
2.11.0

